File: java/minimum_height_trees.java
public class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> res=new ArrayList<Integer>();
        List<Integer> leaf=new ArrayList<Integer>();
        if (n==0) return res;
        if (n==1) {
            res.add(0);
            return res;
        } 
        if (n==2) {
            res.add(0);
            res.add(1);
            return res;
        }
        
        HashSet<Integer>[] graph=new HashSet[n];
        for (int i=0; i<n; i++) {
            graph[i]=new HashSet<Integer>();
        }
        
        for (int i=0; i<edges.length; i++) {
            graph[edges[i][0]].add(edges[i][1]);
            graph[edges[i][1]].add(edges[i][0]);
        }
        
        while (true) {
            res.clear();            
            leaf.clear();
            for (int i=0; i<n; i++) {
                // non-leaf node
                if (graph[i].size()>1) {
                    res.add(i);
                } else if (graph[i].size()==1) {
                    leaf.add(i);
                }
            }

            if (res.size()<=2) return res;
            
            for (int i=0; i<leaf.size(); i++) {
                // leaf node
                for (int neighbor:graph[leaf.get(i)]) {
                    graph[neighbor].remove(leaf.get(i));
                }
                graph[leaf.get(i)].clear();
            }
        }        
    }
}

File: python/minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """        
        mp=collections.defaultdict(set)
        for u, v in edges:
            mp[u].add(v)
            mp[v].add(u)
        
        res=[i for i in range(n) if len(mp[i])<=1]
        while n>2:
            nxt=[]
            for i in res:
                for v in mp[i]:
                    mp[v].remove(i)
                    if len(mp[v])==1: nxt.append(v)
            
            n-=len(res)
            res=nxt
        
        return res
        
